### ISSUE TITLE
.d.ts typing files are out of date

### ISSUE THREAD

**Issue by balajikris** [Azure/azure-sdk-for-node#2083]
I tried writing a "hello world" storage sample in TS and was depending on "azure-arm-resource" and "azure-arm-storage" and I encountered compilation errors from tsc.exe

Here is a sample output:
~~~
E:\repos\samples\storage-node-ts>tsc
node_modules/azure-arm-resource/lib/resource/models/index.d.ts(12,1): error TS1036: Statements are not allowed in ambient contexts.
node_modules/azure-arm-resource/lib/resource/models/index.d.ts(13,34): error TS2339: Property 'CloudError' does not exist on type 'type
of "E:/repos/samples/storage-node-ts/node_modules/ms-rest-azure/lib/index"'.
node_modules/azure-arm-resource/lib/resource/models/index.d.ts(648,35): error TS2304: Cannot find name 'BaseResource'.
node_modules/azure-arm-resource/lib/resource/models/index.d.ts(1086,38): error TS2304: Cannot find name 'BaseResource'.
node_modules/azure-arm-storage/lib/models/index.d.ts(12,1): error TS1036: Statements are not allowed in ambient contexts.
node_modules/azure-arm-storage/lib/models/index.d.ts(13,34): error TS2339: Property 'CloudError' does not exist on type 'typeof "E:/rep
os/samples/storage-node-ts/node_modules/ms-rest-azure/lib/index"'.
node_modules/azure-arm-storage/lib/models/index.d.ts(274,35): error TS2304: Cannot find name 'BaseResource'.
~~~

### This is a running list of things here that need to be updated/fixed:

**azure-arm-resource** is special (generated from composite swagger) and it presents unique problems:
1. No index.d.ts (entry point) in azure-arm-resource
2. If you were to create an index.d.ts and re-export things like so `export * from './resource/resourceManagementClient';` you'd see conflicts as various managementClients export types with same names.

**For all packages:**
1. use `export {name} from` or `export *` syntax instead of using `export =` syntax. `export =` is special, refer to typescript handbook.
> using export = means, corresponding import syntax must be compatible. When importing a module using export =, TypeScript-specific import let = require("module") must be used to import the module.
~~~
    for eg. the 'CloudError' compilation error will go away if we rewrite this line `exports.CloudError = msRestAzure.CloudError;` as `export {CloudError} from 'ms-rest-azure';` in `index.d.ts`
~~~
2. add `"typings": "./lib/index.d.ts"` to package.json - this is necessary if the entry point is not named index.d.ts

3. make parameters optional by appending a ? to their names. without this there is no way to omit parameters in TS, like we can in JS. 
the current generated code does not have optional parameters and if the client passes `null` for those, that changes the semantics of the call -- i.e., the server takes a different code path.
~~~
    for eg. constructor(credentials: ServiceClientCredentials, subscriptionId: string, baseUri: string, options?: ServiceClientOptions);
~~~

4. fix member access by prefixing with type name. 
~~~
    for eg. the last error in the above listing, about `BaseResource` will go away if line 274 of storage\index.d.ts is rewritten to ` msRestAzure.BaseResource`
~~~

5. see all bugs duped to this master bug and ensure they're all fixed.

**Comment by balajikris** [comment]
/cc: @amarzavery @fiveisprime 

the above list is not complete, this is what I encountered while writing a very simple sample (with 2 calls) in TS. Ideally, writing a full sample in TS should expose more generic problems such as above.

**Comment by atrauzzi** [comment]
`@types/azure-sb` is missing some service bus definitions:

![image](https://cloud.githubusercontent.com/assets/115199/23829313/e77545e6-06b4-11e7-87f6-3c7f6b160271.png)

**Comment by atrauzzi** [comment]
I'm also curious, what's the intended usage of this library when it comes to the multiple-sub-packages and `@types` packages?

What combination should I be using?

**Comment by balajikris** [comment]
@atrauzzi - This package is data plane for service bus and is managed by the service bus team. This bug (#2083) tracks .d.ts files generated by github.com/Azure/autorest. So, I'm forking the service bus issue to #2099 and continuing the conversation there.

**Comment by atrauzzi** [comment]
I'm wondering if there's any concrete advice on what I should be using?  I had to switch to using `azure` vs `azure-sb` just to get a successful import of the package in my project.  There may be some issues in the current strategy of maintaining separate packages for azure offerings that need to be tidied up?

**Comment by balajikris** [comment]
tagging @amarzavery and @fiveisprime  as well for the question on package strategy.

**Comment by balajikris** [comment]
Notes:

1. A gap I discovered working on this is that, after making TS related changes to autorest, there is a TS based test suite in autorest that runs. Getting those tests passing seemingly has no correlation to how usable the generated d.ts is. For e.g: passing of tests should imply, i should be able to code against storagemanagement.d.ts and that should should compile and the transpiled javascript should run without any issues. 
But sadly, that doesn't seem to be the case today. While actual serviceclient.d.ts were not even compiling, the tests were regularly passing. We need to bridge this gap and have tests that assert the expected end behavior is indeed happening. 

2. Could not export models from a top level index.d.ts because the corresponding javascript entry point does not export models. It only exports the serviceClient. So, the transpiled js code will never work.
The reason it does so because, it lends to an idiomatic javascript experience where people can say
```js
var client = require('package');
new Client(...);
```
and not have to write code like
```js
var storage = require('package');
new storage.Client(...);
new storage.models.Account(...);
```
as there is no strong typing in javascript. 

The workaround for TS would now be:
```TS
import * as Models from '..\node_modules\package\models\';
```
Documenting this, for future.

**Comment by balajikris** [comment]
This is now complete and shipped with `2.0.0-preview` version of SDK packages.
